### PR TITLE
Fixed bug in function that checks if a path has already been completed

### PR DESCRIPTION
**Pull request by joefitter**
- Path being checked against the list of steps stored in sessionModel included the baseUrl
- Remove base url if found in path to be checked

### CODE REVIEW

**Comment by joefitter** [comment]
@daniel-ac-martin @easternbloc @JoeChapman  could you have a 👀  at this when you get a sec please?

**Comment by daniel-ac-martin** [comment]
@joefitter Don't suppose you fancy adding a test or two to demonstrate the bug? i.e. Tests that use a non-trivial baseUrl. We're clearly lacking coverage in this area.

I think when you remove the baseUrl you should be sure you only remove from the start of the string. I can envisage obscure bugs otherwise. (You could use a regex or remove the first n characters. Regex would probably be safer.)

Nice catch BTW. - Told you it would replace the old bugs with some new ones. (Hopefully fewer though!):-)

**Comment by joefitter** [comment]
@daniel-ac-martin No prob - I have added a couple, and altered the descriptions of some of the existing ones to better describe what they are doing

**Comment by joefitter** [comment]
@JoeChapman @gxxm are you able to merge this when you get a sec so I can rebase against it?

File: test/lib/base-controller.js [hunk 1]
```diff
@@ -329,7 +329,67 @@ describe('lib/base-controller', function () {
...
+              }
+            }];
+            controller.options.continueOnEdit = false;
+            req.params.action = 'edit';
+            req.baseUrl = '/a-base-url';
+            controller.getNextStep(req).should.equal('/a-base-url/confirm');
+          });
+
+          it('should return /a-base-url/target-page if completed check falsy', function () {
+            getStub.returns(['/a-base-url/target-page']);
```
Comment (daniel-ac-martin) [inline test/lib/base-controller.js]: Wouldn't it make more sense for this line to be?:
getStub.returns([]);

(As your point in the PR is that the list of steps do NOT include the baseUrl.)

In fact, does this test (and the one below it) test anything new. Perhaps they should be removed?

@JoeChapman: Thoughts?

File: test/lib/base-controller.js [hunk 2]
```diff
@@ -329,7 +329,67 @@ describe('lib/base-controller', function () {
...
+              }
+            }];
+            controller.options.continueOnEdit = true;
+            req.params.action = 'edit';
+            req.baseUrl = '/a-base-url';
+            controller.getNextStep(req).should.equal('/a-base-url/target-page/edit');
+          });
+
+          it('should not append "edit" to the path if contains function returns falsy', function () {
+            getStub.returns(['/a-base-url/target-page']);
```
Comment (daniel-ac-martin) [inline test/lib/base-controller.js]: See above.

**Comment by daniel-ac-martin** [comment]
Apart from the tests that could do with being cleaned up a bit, this LGTM.

File: test/lib/base-controller.js (hunk 2 above)
Comment (joefitter) [inline test/lib/base-controller.js]: 👍  Yeah on second thoughts, they aren't testing anything new. The bottom one I added has it covered - removing now

**Comment by joefitter** [comment]
Had a conversation with @JoeChapman the other day about potentially extracting these 'helper' functions so they can be individually unit tested - at the moment we are testing a lot of functionality with each test due to the way the controller is written and helpers are defined

File: test/lib/base-controller.js [hunk 3]
```diff
@@ -329,7 +329,37 @@ describe('lib/base-controller', function () {
             }];
             controller.options.continueOnEdit = false;
             req.params.action = 'edit';
-            controller.getNextStep(req).should.contain('/confirm');
+            controller.getNextStep(req).should.equal('/confirm');
+          });
+
+          it('should return /a-base-url/target-page if completed check falsy', function () {
+            getStub.returns(['/a-base-url/target-page']);
```
Comment (daniel-ac-martin) [inline test/lib/base-controller.js]: Did you remove the wrong tests?
This still has the baseUrl in the list of completed steps.

**Comment by daniel-ac-martin** [comment]
It seems a little odd to me to expose functions just for the purposes of unit testing, but I know what you mean.
I'll defer to the other guys on the issue.

BTW I think the tests could be made simpler for free just by replacing the 'condition' function with either true of false. Then there's no need to set req.params.

File: test/lib/base-controller.js (hunk 3 above)
Comment (joefitter) [inline test/lib/base-controller.js]: Yep - that's exactly what I did!

**Comment by joefitter** [comment]
I don't necessarily mean just for unit testing, some helpers could potentially be used elsewhere in the application if the same task/check needs to happen elsewhere - checking if a step has been visited in particular. Could group them together and expose them through `helpers.js` or something along those lines. That way they can be tested and shared

**Comment by joefitter** [comment]
Also the more logic we can abstract from the core controller methods, the better IMO

**Comment by daniel-ac-martin** [comment]
@joefitter: LGTM, ready for me to merge?

**Comment by joefitter** [comment]
@daniel-ac-martin brill, yes please

**Pull request merged**